File: server/app/api/routes/simc.py
import re
from collections.abc import Mapping
from uuid import UUID

from fastapi import APIRouter, Depends, Header
from pydantic import BaseModel, ConfigDict, Field

from server.app.api.dependencies import (
    require_mutating_principal,
    require_principal,
    simulation_application,
)
from server.app.api.errors import ApiProblem
from server.app.identity.domain import Principal
from server.app.simulation.application import (
    SimulationApplication,
    SimulationApplicationError,
    SimulationJobView,
    validated_simulation_result_provenance,
    public_simulation_report,
    public_simulation_metric_error,
)
from server.app.simulation.domain import SimulationAttempt, SourceSnapshot


router = APIRouter(prefix="/api/v2/simc")
_SAFE_CODE = re.compile(r"[A-Za-z0-9_.-]{1,128}\Z")


class SourceSnapshotCreateBody(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    source_url: str = Field(alias="sourceUrl", min_length=1, max_length=2048)


class SimulationCreateBody(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    snapshot_id: UUID = Field(alias="snapshotId")
    scenario: dict[str, object]


def _iso(value: object) -> str:
    return value.isoformat() if hasattr(value, "isoformat") else str(value or "")


def _bounded_codes(value: object, maximum: int = 64) -> list[str]:
    if not isinstance(value, (list, tuple)):
        return []
    return [
        item
        for item in value[:maximum]
        if isinstance(item, str) and _SAFE_CODE.fullmatch(item) is not None
    ]


def _character_payload(snapshot: SourceSnapshot | None) -> dict[str, object] | None:
    raw = snapshot.snapshot.get("character") if snapshot is not None else None
    if not isinstance(raw, Mapping):
        return None
    def text(key):
        value = raw.get(key)
        return value[:160] if isinstance(value, str) else ""
    return {"name": text("name"), "className": text("classKey"),
            "specialization": text("specKey"), "race": text("raceKey"),
            "level": raw.get("level") if type(raw.get("level")) is int and 0 < raw["level"] <= 1000 else None}


def _snapshot_payload(snapshot: SourceSnapshot, workbench: bool = False) -> dict[str, object]:
    raw_snapshot = snapshot.snapshot if isinstance(snapshot.snapshot, Mapping) else {}
    source_revision = str(snapshot.provenance.get("sourceRevision") or "").strip()
    if len(source_revision) > 160:
        source_revision = ""
    return {
        "id": str(snapshot.id),
        **({"character": _character_payload(snapshot)} if workbench else {}),
        "provider": snapshot.provider.value,
        "sourceUrl": snapshot.source_url,
        "sourceKey": snapshot.source_key,
        "revision": snapshot.revision,
        "readiness": snapshot.readiness.value,
        "missingFields": _bounded_codes(raw_snapshot.get("missingFields")),
        "blockers": _bounded_codes(raw_snapshot.get("readinessBlockers")),
        "fetchedAt": _iso(snapshot.fetched_at),
        "provenance": {
            "sourceRevision": source_revision or None,
            "sourceRawSha256": snapshot.raw_sha256,
        },
    }


def _job_summary(view: SimulationJobView, workbench: bool = False) -> dict[str, object]:
    validated_simulation_result_provenance(view)
    job = view.job
    public_error = str(job.public_error_code or "")
    if public_error and _SAFE_CODE.fullmatch(public_error) is None:
        public_error = "SIMC_FAILED"
    return {
        "id": str(job.id),
        **({"character": _character_payload(view.snapshot), "scenario": view.scenario,
            "metric": {"name": view.result.primary_metric_name, "value": view.result.primary_metric_value} if view.result else None} if workbench else {}),
        "snapshotId": str(job.snapshot_id),
        "status": job.status.value,
        "scenarioHash": job.scenario_hash,
        "compilerRevision": job.compiler_revision,
        "runtimeRevision": job.runtime_revision,
        "errorCode": public_error or None,
        "createdAt": _iso(job.created_at),
        "updatedAt": _iso(job.updated_at),
    }


def _attempt_payload(attempt: SimulationAttempt) -> dict[str, object]:
    diagnostic = str(attempt.diagnostic or "")
    if diagnostic == "succeeded":
        diagnostic = "SUCCEEDED"
    elif diagnostic and _SAFE_CODE.fullmatch(diagnostic) is None:
        diagnostic = "SIMC_DIAGNOSTIC_REDACTED"
    return {
        "attemptNumber": attempt.attempt_number,
        "startedAt": _iso(attempt.started_at),
        "finishedAt": _iso(attempt.finished_at) if attempt.finished_at is not None else None,
        "exitCode": attempt.exit_code,
        "diagnosticCode": diagnostic or None,
    }


def _result_payload(view: SimulationJobView, workbench: bool = False) -> dict[str, object] | None:
    provenance = validated_simulation_result_provenance(view)
    result = view.result
    if result is None:
        return None
    if provenance is None:
        raise SimulationApplicationError("SIMC_RESULT_INVALID", "simulation result is invalid")
    return {
        "id": str(result.id),
        "profileSha256": result.profile_sha256,
        **({"report": public_simulation_report(view), "metricError": public_simulation_metric_error(view)} if workbench else {}),
        "metricName": result.primary_metric_name,
        "metricValue": result.primary_metric_value,
        "compilerRevision": result.compiler_revision,
        "runtimeRevision": result.runtime_revision,
        "provenance": provenance,
        "createdAt": _iso(result.created_at),
    }


def _job_detail(view: SimulationJobView, workbench: bool = False) -> dict[str, object]:
    return {
        **_job_summary(view, workbench),
        "attempts": [_attempt_payload(attempt) for attempt in view.attempts[:20]],
        "result": _result_payload(view, workbench),
    }


def _raise_simulation_error(error: SimulationApplicationError) -> None:
    status_code = {
        "INVALID_LINK": 422,
        "INVALID_CURSOR": 422,
        "SIMULATION_NOT_FOUND": 404,
        "SNAPSHOT_NOT_FOUND": 404,
        "SNAPSHOT_NOT_READY": 409,
        "IDEMPOTENCY_CONFLICT": 409,
        "IDEMPOTENCY_KEY_REQUIRED": 422,
        "IDEMPOTENCY_KEY_INVALID": 422,
        "SCENARIO_INVALID": 422,
        "TALENTS_INVALID": 422,
        "PROFILE_TOO_LARGE": 422,
        "COMPILER_UNAVAILABLE": 503,
        "RUNTIME_UNAVAILABLE": 503,
        "SIMC_RESULT_INVALID": 500,
    }.get(error.code, 500)
    raise ApiProblem(status_code=status_code, code=error.code, message=error.message) from error


@router.get("/runtime")
def runtime_info(
    principal: Principal = Depends(require_principal),
    application: SimulationApplication = Depends(simulation_application),
) -> dict[str, object]:
    return application.runtime_info(principal)


@router.post("/snapshots", status_code=201)
def create_snapshot(
    body: SourceSnapshotCreateBody,
    view: str | None = None,
    principal: Principal = Depends(require_mutating_principal),
    application: SimulationApplication = Depends(simulation_application),
) -> dict[str, object]:
    try:
        return _snapshot_payload(application.resolve_source(principal, body.source_url), view == "workbench")
    except SimulationApplicationError as error:
        _raise_simulation_error(error)


@router.get("/snapshots/{snapshot_id}")
def get_snapshot(
    snapshot_id: UUID,
    view: str | None = None,
    principal: Principal = Depends(require_principal),
    application: SimulationApplication = Depends(simulation_application),
) -> dict[str, object]:
    try:
        return _snapshot_payload(application.read_snapshot(principal, snapshot_id), view == "workbench")
    except SimulationApplicationError as error:
        _raise_simulation_error(error)


@router.get("/jobs")
def list_jobs(
    principal: Principal = Depends(require_principal),
    application: SimulationApplication = Depends(simulation_application),
    view: str | None = None,
    cursor: str | None = None,
    limit: int = 20,
) -> dict[str, object]:
    try:
        page = application.list_jobs(principal, cursor, limit)
        return {
            "items": [_job_summary(item, view == "workbench") for item in page.items],
            "nextCursor": page.next_cursor,
        }
    except (SimulationApplicationError, TypeError, ValueError) as error:
        if isinstance(error, SimulationApplicationError):
            _raise_simulation_error(error)
        raise ApiProblem(
            status_code=422,
            code="INVALID_LIMIT",
            message="simulation limit is invalid",
        ) from error


@router.post("/jobs", status_code=202)
def create_job(
    body: SimulationCreateBody,
    view: str | None = None,
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
    principal: Principal = Depends(require_mutating_principal),
    application: SimulationApplication = Depends(simulation_application),
) -> dict[str, object]:
    try:
        job = application.submit(
            principal,
            body.snapshot_id,
            body.scenario,
            idempotency_key or "",
        )
        return _job_detail(application.read_job(principal, job.id), view == "workbench")
    except SimulationApplicationError as error:
        _raise_simulation_error(error)


@router.get("/jobs/{job_id}")
def get_job(
    job_id: UUID,
    view: str | None = None,
    principal: Principal = Depends(require_principal),
    application: SimulationApplication = Depends(simulation_application),
) -> dict[str, object]:
    try:
        return _job_detail(application.read_job(principal, job_id), view == "workbench")
    except SimulationApplicationError as error:
        _raise_simulation_error(error)


__all__ = ["router"]

File: packages/domain/src/simc.ts
import { isSimulationCharacter, isSimulationReport, isSimulationScenario, type SimulationCharacter, type SimulationReport, type SimulationScenario } from './simc-workbench'

export type SourceProvider = 'raiderio' | 'warcraftlogs'
export type SourceReadiness =
  | 'INVALID_LINK'
  | 'CHARACTER_NOT_FOUND'
  | 'ACCESS_RESTRICTED'
  | 'SNAPSHOT_UNAVAILABLE'
  | 'INCOMPLETE_FOR_SIMC'
  | 'READY_FOR_SIMC'

export type SimulationJobStatus = 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled'
export type SimulationMetricName = 'dps' | 'hps'

export interface SourceSnapshotProvenance {
  sourceRevision: string | null
  sourceRawSha256: string
}

export interface SourceSnapshotView {
  character?: SimulationCharacter | null
  id: string
  provider: SourceProvider
  sourceUrl: string
  sourceKey: string
  revision: number
  readiness: SourceReadiness
  missingFields: readonly string[]
  blockers: readonly string[]
  fetchedAt: string
  provenance: SourceSnapshotProvenance
}

export interface SimulationJobSummary {
  character?: SimulationCharacter | null
  scenario?: SimulationScenario | null
  metric?: { name: SimulationMetricName; value: number } | null
  id: string
  snapshotId: string
  status: SimulationJobStatus
  scenarioHash: string
  compilerRevision: string
  runtimeRevision: string
  errorCode: string | null
  createdAt: string
  updatedAt: string
}

export interface SimulationAttemptView {
  attemptNumber: number
  startedAt: string
  finishedAt: string | null
  exitCode: number | null
  diagnosticCode: string | null
}

export interface SimulationResultProvenance {
  snapshotId: string
  sourceRevision: string
  sourceRawSha256: string
  profileSha256: string
  compilerRevision: string
  runtimeRevision: string
  scenarioHash: string
}

export interface SimulationResultView {
  report?: SimulationReport | null
  id: string
  profileSha256: string
  metricName: SimulationMetricName
  metricValue: number
  compilerRevision: string
  runtimeRevision: string
  provenance: SimulationResultProvenance
  createdAt: string
}

export interface SimulationJobDetail extends SimulationJobSummary {
  attempts: readonly SimulationAttemptView[]
  result: SimulationResultView | null
}

export interface SimulationJobPage {
  items: readonly SimulationJobSummary[]
  nextCursor: string | null
}


function record(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function exactKeys(value: Record<string, unknown>, keys: readonly string[], optional: readonly string[] = []): boolean {
  const actual = Object.keys(value)
  return keys.every(key => actual.includes(key)) && actual.every((key) => keys.includes(key) || optional.includes(key))
}

function workbenchSummary(value: Record<string, unknown>): boolean {
  const metric = value['metric']
  return (!('character' in value) || value['character'] === null || isSimulationCharacter(value['character']))
    && (!('scenario' in value) || value['scenario'] === null || isSimulationScenario(value['scenario']))
    && (!('metric' in value) || metric === null || (record(metric) && exactKeys(metric, ['name', 'value']) && (metric['name'] === 'dps' || metric['name'] === 'hps') && typeof metric['value'] === 'number' && Number.isFinite(metric['value']) && metric['value'] > 0))
}

function boundedString(value: unknown, maximum = 160): value is string {
  return typeof value === 'string' && value.length > 0 && value.length <= maximum
}

function isoDate(value: unknown): value is string {
  return boundedString(value, 128) && Number.isFinite(Date.parse(value))
}

function uuid(value: unknown): value is string {
  return typeof value === 'string'
    && /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/iu.test(value)
}

function sha256(value: unknown): value is string {
  return typeof value === 'string' && /^[0-9a-f]{64}$/u.test(value)
}

function positiveInteger(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value > 0
}

function code(value: unknown): value is string {
  return typeof value === 'string' && /^[A-Za-z0-9_.-]{1,128}$/u.test(value)
}

function codeList(value: unknown): value is readonly string[] {
  return Array.isArray(value) && value.length <= 64 && value.every(code)
}

function isSourceSnapshotProvenance(value: unknown): value is SourceSnapshotProvenance {
  if (!record(value) || !exactKeys(value, ['sourceRevision', 'sourceRawSha256'])) return false
  return (value['sourceRevision'] === null || boundedString(value['sourceRevision']))
    && sha256(value['sourceRawSha256'])
}

export function isSourceSnapshotView(value: unknown): value is SourceSnapshotView {
  if (!record(value) || !exactKeys(value, [
    'id',
    'provider',
    'sourceUrl',
    'sourceKey',
    'revision',
    'readiness',
    'missingFields',
    'blockers',
    'fetchedAt',
    'provenance',
  ], ['character'])) return false
  const readiness = value['readiness']
  return (!('character' in value) || value['character'] === null || isSimulationCharacter(value['character']))
    && uuid(value['id'])
    && (value['provider'] === 'raiderio' || value['provider'] === 'warcraftlogs')
    && boundedString(value['sourceUrl'], 2048)
    && /^https:\/\//u.test(value['sourceUrl'])
    && boundedString(value['sourceKey'], 512)
    && positiveInteger(value['revision'])
    && (
      readiness === 'INVALID_LINK'
      || readiness === 'CHARACTER_NOT_FOUND'
      || readiness === 'ACCESS_RESTRICTED'
      || readiness === 'SNAPSHOT_UNAVAILABLE'
      || readiness === 'INCOMPLETE_FOR_SIMC'
      || readiness === 'READY_FOR_SIMC'
    )
    && codeList(value['missingFields'])
    && codeList(value['blockers'])
    && isoDate(value['fetchedAt'])
    && isSourceSnapshotProvenance(value['provenance'])
}

function isSimulationJobSummaryRecord(value: unknown): value is SimulationJobSummary {
  if (!record(value) || !exactKeys(value, [
    'id',
    'snapshotId',
    'status',
    'scenarioHash',
    'compilerRevision',
    'runtimeRevision',
    'errorCode',
    'createdAt',
    'updatedAt',
  ], ['character', 'scenario', 'metric'])) return false
  const status = value['status']
  return workbenchSummary(value) && uuid(value['id'])
    && uuid(value['snapshotId'])
    && (
      status === 'queued'
      || status === 'running'
      || status === 'succeeded'
      || status === 'failed'
      || status === 'cancelled'
    )
    && sha256(value['scenarioHash'])
    && boundedString(value['compilerRevision'])
    && boundedString(value['runtimeRevision'])
    && (value['errorCode'] === null || code(value['errorCode']))
    && isoDate(value['createdAt'])
    && isoDate(value['updatedAt'])
}

function isSimulationAttemptView(value: unknown): value is SimulationAttemptView {
  if (!record(value) || !exactKeys(value, [
    'attemptNumber',
    'startedAt',
    'finishedAt',
    'exitCode',
    'diagnosticCode',
  ])) return false
  return positiveInteger(value['attemptNumber'])
    && isoDate(value['startedAt'])
    && (value['finishedAt'] === null || isoDate(value['finishedAt']))
    && (value['exitCode'] === null || (typeof value['exitCode'] === 'number' && Number.isInteger(value['exitCode'])))
    && (value['diagnosticCode'] === null || code(value['diagnosticCode']))
}

function isSimulationResultProvenance(value: unknown): value is SimulationResultProvenance {
  if (!record(value) || !exactKeys(value, [
    'snapshotId',
    'sourceRevision',
    'sourceRawSha256',
    'profileSha256',
    'compilerRevision',
    'runtimeRevision',
    'scenarioHash',
  ])) return false
  return uuid(value['snapshotId'])
    && boundedString(value['sourceRevision'])
    && sha256(value['sourceRawSha256'])
    && sha256(value['profileSha256'])
    && boundedString(value['compilerRevision'])
    && boundedString(value['runtimeRevision'])
    && sha256(value['scenarioHash'])
}

function isSimulationResultView(value: unknown): value is SimulationResultView {
  if (!record(value) || !exactKeys(value, [
    'id',
    'profileSha256',
    'metricName',
    'metricValue',
    'compilerRevision',
    'runtimeRevision',
    'provenance',
    'createdAt',
  ], ['report'])) return false
  return (!('report' in value) || value['report'] === null || (isSimulationReport(value['report']) && value['report'].metric.name === value['metricName'] && value['report'].metric.value === value['metricValue']))
    && uuid(value['id'])
    && sha256(value['profileSha256'])
    && (value['metricName'] === 'dps' || value['metricName'] === 'hps')
    && typeof value['metricValue'] === 'number'
    && Number.isFinite(value['metricValue'])
    && value['metricValue'] > 0
    && boundedString(value['compilerRevision'])
    && boundedString(value['runtimeRevision'])
    && isSimulationResultProvenance(value['provenance'])
    && isoDate(value['createdAt'])
}

export function isSimulationJobPage(value: unknown): value is SimulationJobPage {
  if (!record(value) || !exactKeys(value, ['items', 'nextCursor'])) return false
  return Array.isArray(value['items'])
    && value['items'].length <= 50
    && value['items'].every(isSimulationJobSummaryRecord)
    && (value['nextCursor'] === null || boundedString(value['nextCursor'], 1024))
}

export function isSimulationJobDetail(value: unknown): value is SimulationJobDetail {
  if (!record(value) || !exactKeys(value, [
    'id',
    'snapshotId',
    'status',
    'scenarioHash',
    'compilerRevision',
    'runtimeRevision',
    'errorCode',
    'createdAt',
    'updatedAt',
    'attempts',
    'result',
  ], ['character', 'scenario', 'metric'])) return false
  if (!workbenchSummary(value)) return false
  const summary: Record<string, unknown> = {
    id: value['id'],
    snapshotId: value['snapshotId'],
    status: value['status'],
    scenarioHash: value['scenarioHash'],
    compilerRevision: value['compilerRevision'],
    runtimeRevision: value['runtimeRevision'],
    errorCode: value['errorCode'],
    createdAt: value['createdAt'],
    updatedAt: value['updatedAt'],
  }
  if (
    !isSimulationJobSummaryRecord(summary)
    || !Array.isArray(value['attempts'])
    || value['attempts'].length > 20
    || !value['attempts'].every(isSimulationAttemptView)
  ) return false
  if (value['result'] === null) return value['status'] !== 'succeeded'
  if (value['status'] !== 'succeeded' || !isSimulationResultView(value['result'])) return false
  const result = value['result']
  return result.compilerRevision === value['compilerRevision']
    && result.runtimeRevision === value['runtimeRevision']
    && result.profileSha256 === result.provenance.profileSha256
    && result.provenance.snapshotId === value['snapshotId']
    && result.provenance.compilerRevision === value['compilerRevision']
    && result.provenance.runtimeRevision === value['runtimeRevision']
    && result.provenance.scenarioHash === value['scenarioHash']
}

export function isSimulationJobSummary(value: unknown): value is SimulationJobSummary {
  return isSimulationJobSummaryRecord(value)
}
